snuba/clickhouse.py

```diff
@@ -375,245 +376,307 @@ def for_schema(self):
-        'sample_expr': sample_expr,
-    }
-
-
-def get_distributed_engine(cluster, database, local_table,
-                           sharding_key=settings.DEFAULT_SHARDING_KEY):
-    return """Distributed(%(cluster)s, %(database)s, %(local_table)s, %(sharding_key)s);""" % {
-        'cluster': cluster,
-        'database': database,
-        'local_table': local_table,
-        'sharding_key': sharding_key,
-    }
-
-
-def get_local_table_definition():
-    return get_table_definition(
-        settings.DEFAULT_LOCAL_TABLE, get_replicated_engine(name=settings.DEFAULT_LOCAL_TABLE)
-    )
-
-
-def get_dist_table_definition():
-    assert settings.CLICKHOUSE_CLUSTER, "CLICKHOUSE_CLUSTER is not set."
-
-    return get_table_definition(
-        settings.DEFAULT_DIST_TABLE,
-        get_distributed_engine(
-            cluster=settings.CLICKHOUSE_CLUSTER,
-            database='default',
-            local_table=settings.DEFAULT_LOCAL_TABLE,
+class TableSchema(object):
+    """
+    Represents the full set of columns in a clickhouse table, with accessors
+    for getting the local and distributed table definitions.
+    """
+
+    def __init__(self):
+        self.CLICKHOUSE_CLUSTER = None
+        self.DATABASE = None
+        self.LOCAL_TABLE = None
+        self.DIST_TABLE = None
+        self.QUERY_TABLE = None
+        self.CAN_DROP = False
+
+        self.SAMPLE_EXPR = None
+        self.ORDER_BY = None
+        self.PARTITION_BY = None
+        self.VERSION_COLUMN = None
+        self.SHARDING_KEY = None
+        self.RETENTION_DAYS = None
+
+        self.ALL_COLUMNS = ColumnSet([])
+
+    def get_local_engine(self):
+        return """
+            ReplicatedReplacingMergeTree('/clickhouse/tables/{shard}/%(name)s', '{replica}', %(version_column)s)
+            PARTITION BY %(partition_by)s
+            ORDER BY %(order_by)s
+            SAMPLE BY %(sample_expr)s;""" % {
+            'name': self.LOCAL_TABLE,
+            'order_by': self.ORDER_BY,
+            'partition_by': self.PARTITION_BY,
+            'version_column': self.VERSION_COLUMN,
+            'sample_expr': self.SAMPLE_EXPR,
+        }
+
+    def get_distributed_engine(self):
+        return """Distributed(%(cluster)s, %(database)s, %(local_table)s, %(sharding_key)s);""" % {
+            'cluster': self.CLICKHOUSE_CLUSTER,
+            'database': self.DATABASE,
+            'local_table': self.LOCAL_TABLE,
+            'sharding_key': self.SHARDING_KEY,
+        }
+
+    def get_table_definition(self, name, engine):
+        return """
+        CREATE TABLE IF NOT EXISTS %(name)s (%(columns)s) ENGINE = %(engine)s""" % {
+            'columns': self.ALL_COLUMNS.for_schema(),
+            'engine': engine,
+            'name': name,
+        }
+
+    def get_local_table_definition(self):
+        return self.get_table_definition(
+            self.LOCAL_TABLE,
+            self.get_local_engine()
         )
-    )
+
+    # TODO what calls this?
```

Comment:
I don't think anything does automatically, but it was created as a helper function so we could use it to create the prod tables in a reproducible way.